File: leetcode/strings-and-array/cpp/334.increasing-triplet-subsequence.cpp
/*
 * @lc app=leetcode id=334 lang=cpp
 *
 * [334] Increasing Triplet Subsequence
 */

// @lc code=start
#include <vector>
#include <string>
#include <set>
#include <algorithm>
#include <iostream>

using namespace std;

/* class Solution {
public:

    bool increasingTriplet(vector<int>& nums) {
		bool	second = false;
		int		second_num;
		int		smallest = nums.front();
		for (auto it = nums.begin() + 1; it != nums.end(); it++) {
			if (second && *it > second_num)
				return true;
			else if (*it < smallest)  // we'll keep second because if it still exists, then it's still good
				smallest = *it;
			else if (!second && *it > smallest) {
				second = true;
				second_num = *it;
			}
			else if (*it > smallest)
				second_num = second_num < *it ? second_num : *it;
		}
		return false;
    }
}; */

class Solution {
	public:

	bool	increasingTriplet(vector<int> &nums) {
		int	min1 = INT_MAX;
		int min2 = INT_MAX;
		
		for (auto num : nums) {
			if (num <= min1) min1 = num;
			else if (num <= min2) min2 = num;
			else return true;
		}
		return false;
	}
} ;

// @lc code=end

File: leetcode/strings-and-array/py/334.increasing-triplet-subsequence.py
#
# @lc app=leetcode id=334 lang=python3
#
# [334] Increasing Triplet Subsequence
#

# @lc code=start
class Solution:
	def increasingTriplet(self, nums: List[int]) -> bool:
		min1 = nums[0]
		min2 = 2**32-1
		for i in nums[1:]:
			if i <= min1: min1 = i
			elif i <= min2: min2 = i
			else: return True
		return False

# @lc code=end

